File: cortapis_app.js
/**
 * Cortapis Security App - JavaScript/Web Version
 * Browser-based encryption with @ marker system for AI indexing protection
 */

class CortapisApp {
  constructor() {
    this.marker = '@';
    this.encryptedTexts = {};
    this.encryptionKey = null;
  }

  /**
   * Generate encryption key from password
   */
  async generateEncryptionKey(password) {
    if (!password) {
      throw new Error('Password required for key generation');
    }

    const encoder = new TextEncoder();
    const data = encoder.encode(password);
    const hashBuffer = await crypto.subtle.digest('SHA-256', data);
    this.encryptionKey = hashBuffer;
    return hashBuffer;
  }

  /**
   * Mark sensitive content with @ symbols in Markdown format
   * Example: admin@example.com -> [@@]admin@example.com
   */
  markSensitiveContent(text, markers = []) {
    let markedText = text;

    // Email pattern: word@domain -> [@@]word@domain
    const emailPattern = /([a-zA-Z0-9._-]+)@([a-zA-Z0-9._-]+)/g;
    markedText = markedText.replace(emailPattern, '[@@]$1@$2');

    // Add markers for custom sensitive content
    markers.forEach(phrase => {
      markedText = markedText.replace(
        new RegExp(phrase, 'g'),
        `[@@]${phrase}`
      );
    });

    return markedText;
  }

  /**
   * Obfuscate text to appear corrupted to alphabetical searches
   * Scatter @ markers throughout to disrupt AI indexing
   */
  obfuscateForIndexing(text) {
    const words = text.split(/\s+/);
    const obfuscatedWords = words.map((word, i) => {
      if (i % 3 === 0 && !word.includes('@')) {
        return `${word}@`;
      }
      return word;
    });

    return obfuscatedWords.join(' ');
  }

  /**
   * Encrypt text using SubtleCrypto API
   */
  async encryptText(text) {
    if (!this.encryptionKey) {
      throw new Error('Encryption key not set');
    }

    const encoder = new TextEncoder();
    const data = encoder.encode(text);
    
    const iv = crypto.getRandomValues(new Uint8Array(12));
    const ciphertext = await crypto.subtle.encrypt(
      {
        name: 'AES-GCM',
        iv: iv
      },
      this.encryptionKey,
      data
    );

    // Combine IV + ciphertext and encode as base64
    const combined = new Uint8Array(iv.length + ciphertext.byteLength);
    combined.set(iv, 0);
    combined.set(new Uint8Array(ciphertext), iv.length);

    return btoa(String.fromCharCode(...combined));
  }

  /**
   * Decrypt text using SubtleCrypto API
   */
  async decryptText(encryptedText) {
    if (!this.encryptionKey) {
      throw new Error('Encryption key not set');
    }

    try {
      const combined = Uint8Array.from(atob(encryptedText), c => c.charCodeAt(0));
      const iv = combined.slice(0, 12);
      const ciphertext = combined.slice(12);

      const plaintext = await crypto.subtle.decrypt(
        {
          name: 'AES-GCM',
          iv: iv
        },
        this.encryptionKey,
        ciphertext
      );

      return new TextDecoder().decode(plaintext);
    } catch (error) {
      throw new Error(`Decryption failed: ${error.message}`);
    }
  }

  /**
   * Remove @ markers (Markdown pseudo-links)
   * Converts "[@@]content" back to "content"
   */
  removeMarkers(markedText) {
    return markedText.replace(/\[@@\]/g, '');
  }

  /**
   * Save file with security layers
   */
  async saveToFile(filename, text, encrypt = true, obfuscate = true) {
    // Step 1: Mark content
    const markedText = this.markSensitiveContent(text);

    // Step 2: Obfuscate if requested
    const obfuscatedText = obfuscate
      ? this.obfuscateForIndexing(markedText)
      : markedText;

    // Step 3: Encrypt if requested
    let encryptedText = obfuscatedText;
    if (encrypt) {
      if (!this.encryptionKey) {
        throw new Error('Encryption key not set');
      }
      encryptedText = await this.encryptText(obfuscatedText);
    }

    // Store in memory
    this.encryptedTexts[filename] = {
      content: encryptedText,
      isEncrypted: encrypt,
      isObfuscated: obfuscate
    };

    return {
      filename,
      marked: markedText,
      obfuscated: obfuscatedText,
      encrypted: encrypt ? encryptedText : null,
      status: 'saved'
    };
  }

  /**
   * Open "call window" (Matrix-style decrypt)
   * Temporarily expose content for viewing/editing
   */
  async openCallWindow(filename) {
    if (!this.encryptedTexts[filename]) {
      throw new Error(`File ${filename} not found`);
    }

    const fileData = this.encryptedTexts[filename];
    let content = fileData.content;

    // Decrypt if encrypted
    if (fileData.isEncrypted) {
      content = await this.decryptText(content);
    }

    // Remove markers
    content = this.removeMarkers(content);

    return content;
  }

  /**
   * Export as Markdown format
   */
  async exportMarkdown(filename) {
    const content = await this.openCallWindow(filename);

    return `# Cortapis Protected Content

## Original Content
\`\`\`markdown
${content}
\`\`\`

## Security Info
- Protected with Cortapis security system
- @ markers prevent AI indexing
- Markdown format maintains readability
- Encrypted for cloud storage

---
*Generated by Cortapis Security App*
`;
  }
}

// Export for use in Node.js or modules
if (typeof module !== 'undefined' && module.exports) {
  module.exports = CortapisApp;
}

File: cortapis_app.py
#!/usr/bin/env python3
"""
Cortapis Security App
Encryption system with @ marker nodes for AI indexing protection
"""

import re
import hashlib
import base64
from typing import List, Tuple, Dict
from cryptography.fernet import Fernet
import json

class CortapisSecurityApp:
    """
    Cortapis: A security solution that encrypts text with @ markers
    to prevent AI indexing and web crawling while maintaining readability.
    
    Features:
    - Mark sensitive content with @ markers
    - Encrypt text before cloud upload
    - Decrypt when opening "call window" (Matrix-style)
    - Markdown-compatible pseudo-links for @ symbols
    """
    
    def __init__(self):
        self.marker = '@'
        self.encryption_key = None
        self.encrypted_texts = {}
        
    def generate_encryption_key(self, password: str = None) -> str:
        """
        Generate or derive encryption key from password
        """
        if password:
            key = base64.urlsafe_b64encode(
                hashlib.sha256(password.encode()).digest()
            )
        else:
            key = Fernet.generate_key()
        
        self.encryption_key = key
        return key.decode()
    
    def mark_sensitive_content(self, text: str, markers: List[str] = None) -> str:
        """
        Mark sensitive content with @ symbols in Markdown pseudo-links format
        Example: "contact [@@]example@.com" becomes marked content
        """
        if markers is None:
            markers = []
        
        marked_text = text
        
        # Replace email-like patterns with @ markers in Markdown format
        # Pattern: word@domain -> [@@]word@domain
        email_pattern = r'([a-zA-Z0-9._-]+)@([a-zA-Z0-9._-]+)'
        marked_text = re.sub(
            email_pattern,
            r'[\@\@]\1@\2',
            marked_text
        )
        
        # Add additional markers for custom sensitive content
        for marker_phrase in markers:
            marked_text = marked_text.replace(
                marker_phrase,
                f'[@@]{marker_phrase}'
            )
        
        return marked_text
    
    def encrypt_text(self, text: str) -> str:
        """
        Encrypt marked text for storage/cloud upload
        """
        if not self.encryption_key:
            self.generate_encryption_key()
        
        cipher = Fernet(self.encryption_key)
        encrypted = cipher.encrypt(text.encode())
        return base64.b64encode(encrypted).decode()
    
    def decrypt_text(self, encrypted_text: str) -> str:
        """
        Decrypt text when opening "call window" (Matrix-style reveal)
        """
        if not self.encryption_key:
            raise ValueError("Encryption key not set. Generate or provide key first.")
        
        try:
            cipher = Fernet(self.encryption_key)
            decoded = base64.b64decode(encrypted_text.encode())
            decrypted = cipher.decrypt(decoded)
            return decrypted.decode()
        except Exception as e:
            raise ValueError(f"Decryption failed: {str(e)}")
    
    def remove_markers(self, marked_text: str) -> str:
        """
        Remove @ markers (pseudo-links format)
        Converts "[@@]content" back to "content"
        """
        # Remove Markdown pseudo-link markers
        cleaned = re.sub(r'\[\@\@\]', '', marked_text)
        return cleaned
    
    def obfuscate_for_indexing(self, text: str) -> str:
        """
        Make text appear corrupted to alphabetical search patterns
        Add @ characters throughout to disrupt indexing
        """
        # Insert @ markers at strategic positions
        words = text.split()
        obfuscated_words = []
        
        for i, word in enumerate(words):
            if i % 3 == 0 and '@' not in word:
                # Scatter @ markers throughout
                obfuscated_words.append(f"{word}@")
            else:
                obfuscated_words.append(word)
        
        return ' '.join(obfuscated_words)
    
    def save_to_file(self, filename: str, text: str, encrypt: bool = True, obfuscate: bool = True) -> Dict:
        """
        Save text with security layers:
        1. Mark sensitive content
        2. Obfuscate (optional)
        3. Encrypt (optional)
        """
        # Step 1: Mark content
        marked_text = self.mark_sensitive_content(text)
        
        # Step 2: Obfuscate if requested
        if obfuscate:
            obfuscated_text = self.obfuscate_for_indexing(marked_text)
        else:
            obfuscated_text = marked_text
        
        # Step 3: Encrypt if requested
        if encrypt:
            if not self.encryption_key:
                self.generate_encryption_key()
            encrypted_text = self.encrypt_text(obfuscated_text)
        else:
            encrypted_text = obfuscated_text
        
        # Store in memory
        self.encrypted_texts[filename] = {
            'content': encrypted_text,
            'is_encrypted': encrypt,
            'is_obfuscated': obfuscate
        }
        
        return {
            'filename': filename,
            'marked': marked_text,
            'obfuscated': obfuscated_text,
            'encrypted': encrypted_text if encrypt else None,
            'status': 'saved'
        }
    
    def open_call_window(self, filename: str) -> str:
        """
        Matrix-style "call window" - decrypt and reveal content
        This is when the file becomes exposed to indexing temporarily
        """
        if filename not in self.encrypted_texts:
            raise ValueError(f"File {filename} not found")
        
        file_data = self.encrypted_texts[filename]
        content = file_data['content']
        
        # Decrypt if encrypted
        if file_data['is_encrypted']:
            content = self.decrypt_text(content)
        
        # Remove obfuscation if applied
        if file_data['is_obfuscated']:
            content = self.remove_markers(content)
        
        # Remove all @ markers
        content = self.remove_markers(content)
        
        return content
    
    def export_markdown(self, filename: str, output_file: str) -> str:
        """
        Export content as Markdown with protection
        Uses Markdown code blocks to contain @ symbols
        """
        content = self.open_call_window(filename)
        
        markdown_content = f"""# Cortapis Protected Content

## Original Content
```markdown
{content}
```

## Security Info
- Protected with Cortapis security system
- @ markers prevent AI indexing
- Markdown format maintains readability
- Encrypted for cloud storage
"""
        
        return markdown_content


def main():
    """
    Example usage of Cortapis Security App
    """
    app = CortapisSecurityApp()
    
    # Generate encryption key with password
    password = "cortapis_secure_2024"
    key = app.generate_encryption_key(password)
    print(f"✓ Encryption key generated\n")
    
    # Example sensitive text
    sensitive_text = """
    Contact me at admin@example.com or support@company.org
    My personal email is user@private.com
    For urgent matters: emergency@contact.net
    """
    
    print("📝 Original text:")
    print(sensitive_text)
    print("\n" + "="*50 + "\n")
    
    # Step 1: Save with security layers
    result = app.save_to_file(
        'sensitive_data.txt',
        sensitive_text,
        encrypt=True,
        obfuscate=True
    )
    
    print("🔒 After marking + obfuscation:")
    print(result['obfuscated'])
    print("\n" + "="*50 + "\n")
    
    print("🔐 Encrypted (safe for cloud):")
    print(result['encrypted'][:80] + "...\n")
    print("\n" + "="*50 + "\n")
    
    # Step 2: Open call window (Matrix-style reveal)
    print("🚪 Opening call window (decrypting)...\n")
    revealed = app.open_call_window('sensitive_data.txt')
    print("📖 Revealed content:")
    print(revealed)
    print("\n" + "="*50 + "\n")
    
    # Step 3: Export as Markdown
    markdown = app.export_markdown('sensitive_data.txt', 'output.md')
    print("📄 Markdown export:")
    print(markdown)


if __name__ == "__main__":
    main()
